File: app/src/main/java/com/jmcintyre/sslshow/SSLInfo.java
/* This file contains code for fetching SSL certificate information
*
* Author: Josh McIntyre
*/
package com.jmcintyre.sslshow;

import android.util.Log;

import java.io.IOException;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

public class SSLInfo
{
    // Connection related Constants
    private final int PORT = 443;
    private final String PROTOCOL = "TLS";
    private final String TAG = "SSLINFO";
    private final String CONN_ERROR = "Unable to create SSL socket for hostname ";

    // Certificate information constants
    private final String SUBJECT_STRING = "\nSubject name: ";
    private final String ISSUER_STRING = "\nIssuer name: ";
    private final String ALGORITHM_STRING = "\nSignature algorithm: ";
    private final String SERIAL_STRING = "\nSerial number: ";

    private final String NOINFO_STRING = "\nNo information available";

    // Connection information
    private Thread sslInfoThread;
    private SSLSocket sslSocket;
    private String hostname;

    // Certificate information
    private X509Certificate certificate;
    private String subjectName;
    private String issuerName;
    private String algorithm;
    private String serialNumber;
    private String issuedOn;
    private String expiresOn;

    // Formatted information consumed by callers

    private String certInfo;


    // This constructor takes the given hostname and fetches the SSL information
    public SSLInfo(String hostname) {
        this.hostname = hostname;
        fetchSSLInfo();
        constructCertInfo();
    }

    // Getter for the certInfo
    public String getCertInfo() {
        return certInfo;
    }

    // Construct the certInfo formatted string
    public void constructCertInfo() {

        // If the certificate is null, return a string indicating that no information could be gathered
        if (certificate == null) {
            certInfo = NOINFO_STRING;
            return;
        }

        // Build up a string-formatted dump of the certificate info
        StringBuilder certInfoBuilder = new StringBuilder();
        certInfoBuilder.append("\nSubject name: ");
        certInfoBuilder.append(subjectName);
        certInfoBuilder.append("\nIssuer name: ");
        certInfoBuilder.append(issuerName);
        certInfoBuilder.append("\nSignature algorithm: ");
        certInfoBuilder.append(algorithm);
        certInfoBuilder.append("\nSerial number: ");
        certInfoBuilder.append(serialNumber);
        certInfoBuilder.append("\nIssued On: ");
        certInfoBuilder.append(issuedOn);
        certInfoBuilder.append("\nExpires on: ");
        certInfoBuilder.append(expiresOn);

        certInfo = certInfoBuilder.toString();

        Log.i(TAG, certInfo);
    }

    // This function requests the SSL connection and gets certificate information
    private void fetchSSLInfo() {

        // Networking operations must be done in their own thread in Android
        // This is to prevent the UI from freezing during a long running network op
        sslInfoThread = new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    SSLContext sslContext = SSLContext.getInstance(PROTOCOL);

                    /* This block initializes a bogus TrustManager that accepts ALL certificates,
                    * including self-signed, revoked, and expired certificates
                    * You should absolutely NEVER do this in production code where the actual validity
                    * of the certificate is important for sending private information over the connection
                    * We never do anything beyond the handshake phase here
                    * The purpose of this utility is to parse and display information about SSL
                    * certificates for education, testing, etc. - so we want this code to work for invalid certs
                    * Without this block, any invalid certificate will throw and exception before we can inspect
                    * the certificates contents and display information
                    */
                    sslContext.init(null, new TrustManager[]{ new X509TrustManager() {

                        private X509Certificate[] accepted;

                        @Override
                        public void checkClientTrusted(X509Certificate[] xcs, String string) throws CertificateException {
                        }

                        @Override
                        public void checkServerTrusted(X509Certificate[] xcs, String string) throws CertificateException {
                            accepted = xcs;
                        }

                        @Override
                        public X509Certificate[] getAcceptedIssuers() {
                            return accepted;
                        }
                    }}, null);

                    // Initialize the socket
                    SSLSocketFactory socketFactory = sslContext.getSocketFactory();
                    sslSocket = (SSLSocket) socketFactory.createSocket(hostname, PORT);

                    // Gather certificate information from the session
                    certificate = (X509Certificate) sslSocket.getSession().getPeerCertificates()[0];

                    subjectName = certificate.getSubjectX500Principal().getName();
                    issuerName = certificate.getIssuerX500Principal().getName();
                    algorithm = certificate.getSigAlgName();
                    serialNumber = certificate.getSerialNumber().toString();
                    issuedOn = certificate.getNotBefore().toString();
                    expiresOn = certificate.getNotAfter().toString();

                } catch (IOException e) {
                    Log.e(TAG, CONN_ERROR + hostname);
                    Log.e(TAG, e.toString());
                } catch (NoSuchAlgorithmException e) {
                    Log.e(TAG, e.toString());
                } catch (KeyManagementException e) {
                    Log.e(TAG, e.toString());
                }
            }
        });

        // Start the SSL information gathering thread
        sslInfoThread.start();

        /* Wait on the thread to finish before exiting anyway as callers need the certInfo to be populated
          This sort of defeats the purpose of running a thread for network ops, but in this context operations are linear anyway
          This satisfies Android's requirement of network ops running in an independent thread while also satisfying
          the need for certInfo to be populated before any callers run getCertInfo()
         */
        try {
            sslInfoThread.join();
        } catch (InterruptedException e) {
            Log.e(TAG, e.toString());
        }
    }
}

File: app_pc/sslinfo.py
# This file contains code for fetching information about an SSL/TLS certificate from a host
#
# Author: Josh McIntyre
#
import cryptography
import cryptography.x509
import socket
import ssl

# Define a class for fetching and parsing certificate information
class SSLInfo:

    PORT = 443
    NO_CERT_MSG = "No certificate information"

    # Initialize the certificate parser with the hostname and fetch information
    def __init__(self, hostname):
    
        # Set the hostname
        self.hostname = hostname
        
        self.cert = None
        self.subject_name = None
        self.issuer_name = None
        self.algorithm = None
        self.serial_number = None
        self.issued_on = None
        self.expires_on = None
    
        # Fetch and parse cert info, then format
        self.fetch_ssl_info()
        self.construct_cert_info()
        
    # Fetch and parse the certificate for a hostname
    def fetch_ssl_info(self):
    
        # Create a connection to fetch the certificate

        try:
            context = ssl.create_default_context()
            context.check_hostname = False
            context.verify_mode = ssl.CERT_NONE
            conn = socket.create_connection((self.hostname, self.PORT))
            sock = context.wrap_socket(conn, server_hostname=self.hostname, )
            sock.settimeout(1)
        
            pem_cert = ssl.DER_cert_to_PEM_cert(sock.getpeercert(True))
        except (socket.gaierror, ConnectionRefusedError):
            self.cert = None
            return
        
        # Get certificate info
        self.cert = cryptography.x509.load_pem_x509_certificate(pem_cert.encode())
        
        self.subject_name = self.cert.subject.rfc4514_string()
        self.issuer_name = self.cert.issuer.rfc4514_string()
        self.algorithm = self.cert.signature_algorithm_oid._name
        self.serial_number = self.cert.serial_number
        self.issued_on = self.cert.not_valid_before
        self.expires_on = self.cert.not_valid_after
    
    # Format the certificate info
    def construct_cert_info(self):
    
        if self.cert:
            self.cert_info = f"Subject name: {self.subject_name}\n"
            self.cert_info += f"Issuer name: {self.issuer_name}\n"
            self.cert_info += f"Algorithm: {self.algorithm}\n"
            self.cert_info += f"Serial number: {self.serial_number}\n"
            self.cert_info += f"Issued on: {self.issued_on}\n"
            self.cert_info += f"Expires on: {self.expires_on}\n"
        else:
            self.cert_info = self.NO_CERT_MSG
